pkg/controller/pubsub/pubsub_config_controller.go

```diff
@@ -0,0 +1,132 @@
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	logf "sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+)
+
+var log = logf.Log.WithName("controller").WithValues(logging.Process, "pubsub_controller")
+
+type Adder struct {
+	PubsubSystem *pubsub.System
+}
+
+func (a *Adder) Add(mgr manager.Manager) error {
+	if len(psClient.Providers()) == 0 {
+		return nil
+	}
+	r := newReconciler(mgr, a.PubsubSystem)
+	return add(mgr, r)
+}
+
+func (a *Adder) InjectOpa(_ *constraintclient.Client) {}
+
+func (a *Adder) InjectWatchManager(_ *watch.Manager) {}
+
+func (a *Adder) InjectControllerSwitch(_ *watch.ControllerSwitch) {}
+
+func (a *Adder) InjectTracker(_ *readiness.Tracker) {}
+
+func (a *Adder) InjectMutationSystem(_ *mutation.System) {}
+
+func (a *Adder) InjectExpansionSystem(_ *expansion.System) {}
+
+func (a *Adder) InjectProviderCache(_ *externaldata.ProviderCache) {}
+
+func (a *Adder) InjectPubsubSystem(pubsubSystem *pubsub.System) {
+	a.PubsubSystem = pubsubSystem
+}
+
+type Reconciler struct {
+	client.Client
+	scheme *runtime.Scheme
+	system *pubsub.System
+}
+
+func newReconciler(mgr manager.Manager, system *pubsub.System) *Reconciler {
+	return &Reconciler{
+		Client: mgr.GetClient(),
+		scheme: mgr.GetScheme(),
+		system: system,
+	}
+}
+
+func add(mgr manager.Manager, r reconcile.Reconciler) error {
+	c, err := controller.New("pusbub-config-controller", mgr, controller.Options{Reconciler: r})
+	if err != nil {
+		return err
+	}
+
+	return c.Watch(
+		&source.Kind{Type: &corev1.ConfigMap{}},
+		&handler.EnqueueRequestForObject{})
+}
+
+func (r *Reconciler) Reconcile(ctx context.Context, request reconcile.Request) (reconcile.Result, error) {
+	log.Info("Reconcile", "request", request, "namespace", request.Namespace, "name", request.Name)
+
+	deleted := false
+	cfg := &corev1.ConfigMap{}
+	err := r.Get(ctx, request.NamespacedName, cfg)
+	if err != nil {
+		if !errors.IsNotFound(err) {
+			return reconcile.Result{}, err
+		}
+		deleted = true
+	}
+
+	if deleted {
+		err := r.system.CloseConnection(request.Name)
+		if err != nil {
+			return reconcile.Result{}, err
+		}
+		log.Info("removed connection", "name", request.Name)
+	} else {
+		providers := psClient.Providers()
+		if len(providers) == 0 {
+			log.Info("No pub sub tool is enabled")
+			return reconcile.Result{}, nil
+		}
+
+		if err != nil {
+			return reconcile.Result{}, err
+		}
+		if len(cfg.Data) == 0 {
+			return reconcile.Result{}, fmt.Errorf(fmt.Sprintf("data missing in configmap %s, unable to configure respective pubsub", request.NamespacedName))
+		}
+
+		if newConnection, ok := providers[cfg.Data["provider"]]; ok {
+			var config interface{}
+			err := json.Unmarshal([]byte(cfg.Data["config"]), &config)
+			if err != nil {
+				return reconcile.Result{}, err
+			}
+			err = r.system.InitiateConnection(ctx, newConnection, config, request.Name)
```

Comment:
What happens if this is the second time reconciling the same config?

For context: k8s objects are semi-regularly re-reconciled, also a user may take an old config map and `kubectl edit` it.

Currently it looks like a second connection with the same name will be created, which will clobber the old connection.

If there is no change in config, do we want the overhead of re-establishing a connection?

Should the old connection be properly torn down?

IMO it's probably best for `system` to handle these details (maybe delegate to driver to determine if the config difference is such that a new connection should be established).

This also implies that `InitiateConnection()` should be less descriptive... `UpsertConnection()`? That copies language we use elsewhere.

Reply:
Added comprehensive handling of connections on reconciles. Also added predicates to prevent reconciling on the configs that are created outside of the installation namespace. This will mandate users to create respective config within the installation namespace. If we don't want this then I think we should come up with an annotation or label that indicates the intent of the config that is being considered for reconciliation. In other words, identification of some sort allows us not to reconcile on configMaps that are not intended for pub-sub config.